File: clients/ios/Lextures/Core/LMS/OrgBrandingAdminLogic.swift
import Foundation
import SwiftUI

/// Org branding, AI governance, and AI provider admin helpers (M14.5).
enum OrgBrandingAdminLogic {
    static let rbacManagePermission = "global:app:rbac:manage"
    static let orgUnitsAdminPermission = "tenant:org:units:admin"
    /// Matches web `PLATFORM_SECRET_PLACEHOLDER` and server `placeholderSecretResponse`.
    static let secretPlaceholder = "••••••••••••"
    static let defaultPrimaryColor = "#4F46E5"
    static let defaultSecondaryColor = "#7C3AED"
    static let defaultProvider = "openrouter"
    static let defaultModelAlias = "claude-3-5-sonnet"
    static let contrastAAThreshold = 4.5

    static let featureKeys: [(key: String, labelKey: String)] = [
        ("ai_tutor", "mobile.admin.orgBranding.ai.feature.aiTutor"),
        ("rag_notebook", "mobile.admin.orgBranding.ai.feature.notebook"),
        ("syllabus_generation", "mobile.admin.orgBranding.ai.feature.syllabus"),
        ("translation", "mobile.admin.orgBranding.ai.feature.translation"),
        ("quiz_generation", "mobile.admin.orgBranding.ai.feature.quiz"),
        ("lesson_generation", "mobile.admin.orgBranding.ai.feature.lesson"),
    ]

    private static let providerLabelKeys: [String: String] = [
        "openrouter": "mobile.admin.orgBranding.provider.openrouter",
        "anthropic": "mobile.admin.orgBranding.provider.anthropic",
        "openai": "mobile.admin.orgBranding.provider.openai",
        "azure_openai": "mobile.admin.orgBranding.provider.azureOpenai",
        "bedrock": "mobile.admin.orgBranding.provider.bedrock",
        "vertex": "mobile.admin.orgBranding.provider.vertex",
    ]

    static func adminSettingsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffMobileAdminSettings
    }

    static func canManage(permissions: [String]) -> Bool {
        permissions.contains(rbacManagePermission) || permissions.contains(orgUnitsAdminPermission)
    }

    static func shouldShowEntry(
        features: MobilePlatformFeatures,
        permissions: [String]
    ) -> Bool {
        adminSettingsEnabled(features) && canManage(permissions: permissions)
    }

    static func canView(
        features: MobilePlatformFeatures,
        permissions: [String]
    ) -> Bool {
        shouldShowEntry(features: features, permissions: permissions)
    }

    static func webBrandingPath() -> String { "/settings/org-branding" }

    static func resolveOrgId(accessToken: String?, courses: [CourseSummary]) -> String? {
        CourseCreateLogic.resolveOrgId(accessToken: accessToken, courses: courses)
    }

    static func resolveAssetURL(_ pathOrURL: String?) -> URL? {
        guard let raw = pathOrURL?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
            return nil
        }
        if raw.hasPrefix("http://") || raw.hasPrefix("https://") {
            return URL(string: raw)
        }
        let path = raw.hasPrefix("/") ? raw : "/\(raw)"
        return AppConfiguration.apiURL(path: path)
    }

    static func isValidHexColor(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.hasPrefix("#") else { return false }
        let hex = String(trimmed.dropFirst())
        guard hex.count == 3 || hex.count == 6 else { return false }
        return hex.allSatisfy { $0.isHexDigit }
    }

    static func normalizeHexColor(_ value: String, fallback: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isValidHexColor(trimmed) else { return fallback }
        let hex = String(trimmed.dropFirst())
        if hex.count == 3 {
            let expanded = hex.map { "\($0)\($0)" }.joined()
            return "#\(expanded.uppercased())"
        }
        return "#\(hex.uppercased())"
    }

    static func color(fromHex value: String) -> Color? {
        let normalized = normalizeHexColor(value, fallback: "")
        guard normalized.hasPrefix("#"), normalized.count == 7 else { return nil }
        let hex = String(normalized.dropFirst())
        guard let int = UInt64(hex, radix: 16) else { return nil }
        let red = Double((int >> 16) & 0xFF) / 255
        let green = Double((int >> 8) & 0xFF) / 255
        let blue = Double(int & 0xFF) / 255
        return Color(red: red, green: green, blue: blue)
    }

    /// Relative luminance contrast ratio of a hex color against white (WCAG).
    static func contrastRatioAgainstWhite(_ hex: String) -> Double? {
        let normalized = normalizeHexColor(hex, fallback: "")
        guard normalized.hasPrefix("#"), normalized.count == 7 else { return nil }
        let raw = String(normalized.dropFirst())
        guard let int = UInt64(raw, radix: 16) else { return nil }
        func channel(_ value: UInt64) -> Double {
            let component = Double(value) / 255
            return component <= 0.03928
                ? component / 12.92
                : pow((component + 0.055) / 1.055, 2.4)
        }
        let red = channel((int >> 16) & 0xFF)
        let green = channel((int >> 8) & 0xFF)
        let blue = channel(int & 0xFF)
        let whiteLuminance = 1.0
        let colorLuminance = 0.2126 * red + 0.7152 * green + 0.0722 * blue
        let lighter = max(whiteLuminance, colorLuminance)
        let darker = min(whiteLuminance, colorLuminance)
        return (lighter + 0.05) / (darker + 0.05)
    }

    static func hasContrastWarning(
        primaryColor: String,
        serverWarning: Bool,
        serverRatio: Double?
    ) -> Bool {
        if serverWarning { return true }
        if let serverRatio, serverRatio < contrastAAThreshold { return true }
        if let local = contrastRatioAgainstWhite(primaryColor), local < contrastAAThreshold {
            return true
        }
        return false
    }

    static func brandingPutBody(
        logoUrl: String?,
        faviconUrl: String?,
        primaryColor: String,
        secondaryColor: String,
        customEmailDisplayName: String?
    ) -> OrgBrandingPutRequest {
        let email = customEmailDisplayName?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return OrgBrandingPutRequest(
            logoUrl: logoUrl,
            faviconUrl: faviconUrl,
            primaryColor: normalizeHexColor(primaryColor, fallback: defaultPrimaryColor),
            secondaryColor: normalizeHexColor(secondaryColor, fallback: defaultSecondaryColor),
            customDomain: nil,
            customEmailDisplayName: (email?.isEmpty == false) ? email : nil
        )
    }

    static func isFeatureEnabled(_ map: [String: Bool], key: String) -> Bool {
        map[key] != false
    }

    static func featuresEnabledPayload(_ enabled: [String: Bool]) -> [String: Bool] {
        var out: [String: Bool] = [:]
        for item in featureKeys {
            out[item.key] = isFeatureEnabled(enabled, key: item.key)
        }
        return out
    }

    static func parseAllowedModels(_ text: String) -> [String]? {
        let models = text
            .split { $0 == "\n" || $0 == "," }
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        return models.isEmpty ? nil : models
    }

    static func allowedModelsText(_ models: [String]?) -> String {
        (models ?? []).joined(separator: "\n")
    }

    static func aiConfigPutBody(
        enabled: [String: Bool],
        allowedModelsText: String
    ) -> AIGovernancePutRequest {
        AIGovernancePutRequest(
            featuresEnabled: featuresEnabledPayload(enabled),
            allowedModels: parseAllowedModels(allowedModelsText)
        )
    }

    /// Returns the BYOK key to send, or nil when the field is empty / still the mask.
    static func byokKeyForSave(_ entered: String) -> String? {
        let trimmed = entered.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty || trimmed == secretPlaceholder {
            return nil
        }
        return trimmed
    }

    static func displaySecretField(byokConfigured: Bool) -> String {
        byokConfigured ? secretPlaceholder : ""
    }

    static func isSecretPlaceholder(_ value: String) -> Bool {
        value.trimmingCharacters(in: .whitespacesAndNewlines) == secretPlaceholder
    }

    static func providerLabel(for provider: String) -> String {
        if let key = providerLabelKeys[provider] {
            return L.dynamicText(key)
        }
        return provider
    }

    static func providerOptions(from settings: AIProviderSettings?) -> [String] {
        let list = settings?.providers ?? Array(providerLabelKeys.keys).sorted()
        return list.isEmpty ? [defaultProvider] : list
    }

    static func modelAliasOptions(from settings: AIProviderSettings?) -> [String] {
        let list = settings?.modelAliases ?? [defaultModelAlias, "gpt-4o", "gemini-1.5-pro"]
        return list.isEmpty ? [defaultModelAlias] : list
    }

    static func aiProviderPutBody(
        provider: String,
        modelAlias: String,
        fallbackProvider: String,
        byokKey: String
    ) -> AIProviderSettingsPutRequest {
        let fallback = fallbackProvider.trimmingCharacters(in: .whitespacesAndNewlines)
        return AIProviderSettingsPutRequest(
            provider: provider.isEmpty ? defaultProvider : provider,
            modelAlias: modelAlias.isEmpty ? defaultModelAlias : modelAlias,
            fallbackProvider: fallback.isEmpty ? nil : fallback,
            byokApiKey: byokKeyForSave(byokKey)
        )
    }

    static func userFacingError(_ error: Error) -> String {
        if let apiError = error as? APIError, case let .httpStatus(_, message) = apiError,
           let message, !message.isEmpty {
            return message
        }
        return L.text("mobile.admin.orgBranding.error")
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/OrgBrandingAdminLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.core.network.ApiError
import kotlin.math.pow

/** Org branding, AI governance, and AI provider admin helpers (M14.5). */
object OrgBrandingAdminLogic {
    const val RBAC_MANAGE_PERMISSION = "global:app:rbac:manage"
    const val ORG_UNITS_ADMIN_PERMISSION = "tenant:org:units:admin"

    /** Matches web `PLATFORM_SECRET_PLACEHOLDER` and server `placeholderSecretResponse`. */
    const val SECRET_PLACEHOLDER = "••••••••••••"
    const val DEFAULT_PRIMARY_COLOR = "#4F46E5"
    const val DEFAULT_SECONDARY_COLOR = "#7C3AED"
    const val DEFAULT_PROVIDER = "openrouter"
    const val DEFAULT_MODEL_ALIAS = "claude-3-5-sonnet"
    const val CONTRAST_AA_THRESHOLD = 4.5

    data class FeatureKey(val key: String, val labelResName: String)

    val FEATURE_KEYS: List<FeatureKey> = listOf(
        FeatureKey("ai_tutor", "mobile_admin_orgBranding_ai_feature_aiTutor"),
        FeatureKey("rag_notebook", "mobile_admin_orgBranding_ai_feature_notebook"),
        FeatureKey("syllabus_generation", "mobile_admin_orgBranding_ai_feature_syllabus"),
        FeatureKey("translation", "mobile_admin_orgBranding_ai_feature_translation"),
        FeatureKey("quiz_generation", "mobile_admin_orgBranding_ai_feature_quiz"),
        FeatureKey("lesson_generation", "mobile_admin_orgBranding_ai_feature_lesson"),
    )

    private val providerLabelKeys: Map<String, String> = mapOf(
        "openrouter" to "mobile_admin_orgBranding_provider_openrouter",
        "anthropic" to "mobile_admin_orgBranding_provider_anthropic",
        "openai" to "mobile_admin_orgBranding_provider_openai",
        "azure_openai" to "mobile_admin_orgBranding_provider_azureOpenai",
        "bedrock" to "mobile_admin_orgBranding_provider_bedrock",
        "vertex" to "mobile_admin_orgBranding_provider_vertex",
    )

    fun adminSettingsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffMobileAdminSettings

    fun canManage(permissions: List<String>): Boolean =
        permissions.contains(RBAC_MANAGE_PERMISSION) ||
            permissions.contains(ORG_UNITS_ADMIN_PERMISSION)

    fun shouldShowEntry(
        features: MobilePlatformFeatures,
        permissions: List<String>,
    ): Boolean = adminSettingsEnabled(features) && canManage(permissions)

    fun canView(
        features: MobilePlatformFeatures,
        permissions: List<String>,
    ): Boolean = shouldShowEntry(features, permissions)

    fun webBrandingPath(): String = "/settings/org-branding"

    fun resolveOrgId(accessToken: String?, courses: List<CourseSummary>): String? =
        CourseCreateLogic.resolveOrgId(accessToken, courses)

    fun isValidHexColor(value: String): Boolean {
        val trimmed = value.trim()
        if (!trimmed.startsWith("#")) return false
        val hex = trimmed.drop(1)
        if (hex.length != 3 && hex.length != 6) return false
        return hex.all { it.isDigit() || it.lowercaseChar() in 'a'..'f' }
    }

    fun normalizeHexColor(value: String, fallback: String): String {
        val trimmed = value.trim()
        if (!isValidHexColor(trimmed)) return fallback
        val hex = trimmed.drop(1)
        return if (hex.length == 3) {
            "#" + hex.map { "$it$it" }.joinToString("").uppercase()
        } else {
            "#${hex.uppercase()}"
        }
    }

    /** Relative luminance contrast ratio of a hex color against white (WCAG). */
    fun contrastRatioAgainstWhite(hex: String): Double? {
        val normalized = normalizeHexColor(hex, "")
        if (!normalized.startsWith("#") || normalized.length != 7) return null
        val raw = normalized.drop(1).toLongOrNull(16) ?: return null
        fun channel(value: Long): Double {
            val c = value / 255.0
            return if (c <= 0.03928) c / 12.92 else ((c + 0.055) / 1.055).pow(2.4)
        }
        val r = channel((raw shr 16) and 0xFF)
        val g = channel((raw shr 8) and 0xFF)
        val b = channel(raw and 0xFF)
        val l1 = 1.0
        val l2 = 0.2126 * r + 0.7152 * g + 0.0722 * b
        val lighter = maxOf(l1, l2)
        val darker = minOf(l1, l2)
        return (lighter + 0.05) / (darker + 0.05)
    }

    fun hasContrastWarning(
        primaryColor: String,
        serverWarning: Boolean,
        serverRatio: Double?,
    ): Boolean {
        if (serverWarning) return true
        if (serverRatio != null && serverRatio < CONTRAST_AA_THRESHOLD) return true
        val local = contrastRatioAgainstWhite(primaryColor)
        return local != null && local < CONTRAST_AA_THRESHOLD
    }

    fun brandingPutBody(
        logoUrl: String?,
        faviconUrl: String?,
        primaryColor: String,
        secondaryColor: String,
        customEmailDisplayName: String?,
    ): OrgBrandingPutRequest {
        val email = customEmailDisplayName?.trim()
        return OrgBrandingPutRequest(
            logoUrl = logoUrl,
            faviconUrl = faviconUrl,
            primaryColor = normalizeHexColor(primaryColor, DEFAULT_PRIMARY_COLOR),
            secondaryColor = normalizeHexColor(secondaryColor, DEFAULT_SECONDARY_COLOR),
            customDomain = null,
            customEmailDisplayName = email?.takeIf { it.isNotEmpty() },
        )
    }

    fun isFeatureEnabled(map: Map<String, Boolean>, key: String): Boolean =
        map[key] != false

    fun featuresEnabledPayload(enabled: Map<String, Boolean>): Map<String, Boolean> =
        FEATURE_KEYS.associate { it.key to isFeatureEnabled(enabled, it.key) }

    fun parseAllowedModels(text: String): List<String>? {
        val models = text
            .split('\n', ',')
            .map { it.trim() }
            .filter { it.isNotEmpty() }
        return models.takeIf { it.isNotEmpty() }
    }

    fun allowedModelsText(models: List<String>?): String =
        models.orEmpty().joinToString("\n")

    fun aiConfigPutBody(
        enabled: Map<String, Boolean>,
        allowedModelsText: String,
    ): AIGovernancePutRequest =
        AIGovernancePutRequest(
            featuresEnabled = featuresEnabledPayload(enabled),
            allowedModels = parseAllowedModels(allowedModelsText),
        )

    /** Returns the BYOK key to send, or null when empty / still the mask. */
    fun byokKeyForSave(entered: String): String? {
        val trimmed = entered.trim()
        if (trimmed.isEmpty() || trimmed == SECRET_PLACEHOLDER) return null
        return trimmed
    }

    fun displaySecretField(byokConfigured: Boolean): String =
        if (byokConfigured) SECRET_PLACEHOLDER else ""

    fun isSecretPlaceholder(value: String): Boolean =
        value.trim() == SECRET_PLACEHOLDER

    fun providerLabelKey(provider: String): String? = providerLabelKeys[provider]

    fun providerOptions(settings: AIProviderSettings?): List<String> {
        val list = settings?.providers ?: providerLabelKeys.keys.sorted()
        return list.ifEmpty { listOf(DEFAULT_PROVIDER) }
    }

    fun modelAliasOptions(settings: AIProviderSettings?): List<String> {
        val list = settings?.modelAliases
            ?: listOf(DEFAULT_MODEL_ALIAS, "gpt-4o", "gemini-1.5-pro")
        return list.ifEmpty { listOf(DEFAULT_MODEL_ALIAS) }
    }

    fun aiProviderPutBody(
        provider: String,
        modelAlias: String,
        fallbackProvider: String,
        byokKey: String,
    ): AIProviderSettingsPutRequest {
        val fallback = fallbackProvider.trim()
        return AIProviderSettingsPutRequest(
            provider = provider.ifBlank { DEFAULT_PROVIDER },
            modelAlias = modelAlias.ifBlank { DEFAULT_MODEL_ALIAS },
            fallbackProvider = fallback.takeIf { it.isNotEmpty() },
            byokApiKey = byokKeyForSave(byokKey),
        )
    }

    fun userFacingError(error: Throwable, genericMessage: String): String =
        (error as? ApiError.HttpStatus)?.message?.takeIf { it.isNotEmpty() } ?: genericMessage
}
